### DISCUSSION TITLE
Suggestion: Use websockets for suggestions

### DISCUSSION THREAD

**Discussion by KTibow** [community/community#10844] [Copilot Conversations]
From what I can tell, Copilot makes a new request (to `https://copilot-proxy.githubusercontent.com/v1/engines/copilot-aatest/completions`) every time a completion is attempted. This means it takes time to set up a new request.

[Websockets](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API) would mean that Copilot could be faster, because it doesn't send headers, and it doesn't establish a new connection. It isn't that hard to implement them from what I understand.

**Comment by davecheney** [comment]
We use HTTP/2 which multiplexes streams over a nice warm TCP flow that lives for the duration of your editing session.

**Reply by KTibow** [reply]
I can see how that would help a lot. However, I think there might be some other ways websockets *could* be faster:
- Sending only part of the updated code (very minor difference, but it *might* help)
- Sending only one request, and receiving alternative solutions

I found a [couple](https://blog.feathersjs.com/http-vs-websockets-a-performance-comparison-da2533f13a77) of [articles](https://medium.com/platform-engineer/web-api-design-35df8167460) which attempted to benchmark them. The first one doesn't say anything about keep-alive, and the second one seems to be more oriented around the server, however websockets seem to generally be faster.

Anyway, I understand it would be a big engineering change, for a possibly small benefit.

**Comment by Chiroyce1** [comment]
I frequently seem to be getting the "Copilot is encountering temporary issues" message - is it because of a network issue or a problem with copilot itself? I do think using WebSockets would make copilot faster and more responsive.

**Reply by davecheney** [reply]
Sadly that error is, in error. I think there is a fix coming that will make complain less.

**Reply by Chiroyce1** [reply]
Cool, thanks.
